UNL-CSE-AMBASSADORS/VITA
# Appointment Notes

This PR enables a volunteer to add notes to appointments and view those notes on the private queue or the appointment management page. Additionally, this **removes** the `notCompletedDescription` field from the `ServicedAppointment` table. This also updates the documents that an admin can download so that it no longer includes the notCompletedDescription data. This also adds a `cancelled` field to ServicedAppointment in place of the notCompletedDescription.

Also, @CollinsSpencer I would really like comments on how I display the notes. Not sure if just appending the author / timestamp to the note text is the correct approach. 

**TESTING ADDING NOTES**
1. Run the table creation and test data script
2. Log in as a preparer (no specific reason to be a preparer, but we under-test as a preparer, so just to make sure)
3. Navigate to the private queue
4. Select an appointment
    * You should see the note section which includes a textarea to add a note
5. Try to hit the "Add" button, you should not be able to
6. Type a note in the text area
7. Hit the "Add" button
    * The note should be added successfully
    * The note should appear in the area above and show that it was created by you a minute ago
8. Add another note
9. Copy the first name of the appointment so we can search it on the appointment management page
10. Choose another appointment on the private queue
    * You should not see the notes you put on the other appointment
    * Add a note to this appointment. Ensure it was successful
11. Navigate to the appointment management page
12. Search for the appointment you added notes to on the private queue
13. Ensure that you see the notes that you added
14. Add a note here. Ensure it was successful


**TESTING DOCUMENTS**
1. Please do this testing AFTER you did the above testing
2. Log in as the site admin
3. Go to the private queue and complete an appointment, be sure to mark any of the filing status options
4. Navigate to the "Print Documents" page
5. Select the current date and all sites. Download the "Appointment Filing Statuses" document
6. Open the Excel file
    * You should see the notes you added in the columns furthest to the fight
    * The filing statuses should be marked properly.

**MISCELLANEOUS TESTING**
I refactored some files and changed a lot of files due to removing that `notCompletedDescription` field. Could you please also test these functionalities:
* Rescheduling an appointment


**DEPLOYMENT STEPS**
The following steps will have to be taken once this is deployed to production.
1. Run the table creation SQL script
2. Run the cancelled updated SQL script
3. Run the /server/temp/DatabaseEvolution.php script **Please note that only Matthew will be able to run this**

**ENSURE THAT THE NOTES HAVE BEEN PROPERLY INSERTED AND THAT THE CANCELLED FIELD HAS BEEN SET CORRECTLY. ONLY PROCEED IF THAT HAS HAPPENED CORRECTLY**

4. Check again and make sure you are certain it is working
5. Run the notCompletedDescription update SQL Script 
6. Remove the /server/temp directory from prod and from master.

**TABLE CREATION SCRIPT FOR PROD**
```
CREATE TABLE Note (
    noteId INTEGER UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
    createdAt DATETIME NOT NULL DEFAULT NOW(),
    note VARCHAR(1000) NOT NULL,
    appointmentId INTEGER UNSIGNED NOT NULL,
    FOREIGN KEY(appointmentId) REFERENCES Appointment(appointmentId),
    createdBy INTEGER UNSIGNED NOT NULL,
    FOREIGN KEY(createdBy) REFERENCES User(userId)
);
```

**CANCELLED UPDATED SQL SCRIPT**
```
ALTER TABLE ServicedAppointment ADD COLUMN cancelled BOOLEAN NOT NULL DEFAULT FALSE;
```

**UPDATE SQL SCRIPT FOR PROD**
```
ALTER TABLE ServicedAppointment DROP COLUMN notCompletedDescription;
```

**NEW DATABASE ERD**
![databaseerd](https://user-images.githubusercontent.com/9295744/40465651-b13f37ae-5ed6-11e8-8f0a-447b32c1fdc1.png)
